File: 85-maximal-rectangle.py
from collections import deque


class Solution:
    def maximalRectangle(self, matrix: list[list[str]]) -> int:
        '''
        We will have a loop that goes through each row. At each row, we know the height we can have for each column
        We can use a monotonically increasing stack to store the index of the possible rectangle and the height at the column
        '''
        n = len(matrix[0])
        max_area = 0
        heights = [0] * (n+1)
        for row in matrix:
            for col_i, col in enumerate(row):
                if col == '0':
                    heights[col_i] = 0
                else:
                    heights[col_i] += 1

            stack = deque()
            for i, h in enumerate(heights):
                prev_i = i
                while stack and stack[-1][1] > h:
                    top_i, top_h = stack.pop()
                    max_area = max(max_area, (i - top_i) * top_h)
                    prev_i = top_i
                stack.append([prev_i, h])

            for i, h in stack:
                max_area = max(max_area, (n - i) * h)

        return max_area


sol = Solution()
print(
    f'output: {sol.maximalRectangle([["1", "0", "1", "0", "0"], ["1", "0", "1", "1", "1"], ["1", "1", "1", "1", "1"], ["1", "0", "0", "1", "0"]])}, expected: 6')
print(f'output: {sol.maximalRectangle([["0"]])}, expected: 0')
print(f'output: {sol.maximalRectangle([["1"]])}, expected: 1')

File: 85-maximal-rectangle.go
//go:build ignore

package main

import "fmt"

func maximalRectangle(matrix [][]byte) int {
	n := len(matrix[0])
	maxArea := 0
	heights := make([]int, n+1)
	for _, row := range matrix {
		for col_i, col := range row {
			if col == '0' {
				heights[col_i] = 0
			} else {
				heights[col_i]++
			}
		}

		stack := make([][]int, 0)
		for i, h := range heights {
			prevI := i
			for len(stack) > 0 && (stack[len(stack)-1][1] > h) {
				top := stack[len(stack)-1]
				maxArea = max(maxArea, (i-top[0])*top[1])
				prevI = top[0]
				stack = stack[:len(stack)-1]
			}
			stack = append(stack, []int{prevI, h})
		}

		for _, top := range stack {
			i, h := top[0], top[1]
			maxArea = max(maxArea, (n-i)*h)
		}
	}
	return maxArea
}

func main() {
	fmt.Printf("Output: %v, expected: 6\n", maximalRectangle([][]byte{{'1', '0', '1', '0', '0'}, {'1', '0', '1', '1', '1'}, {'1', '1', '1', '1', '1'}, {'1', '0', '0', '1', '0'}}))
	fmt.Printf("Output: %v, expected: 0\n", maximalRectangle([][]byte{{'0'}}))
	fmt.Printf("Output: %v, expected: 1\n", maximalRectangle([][]byte{{'1'}}))
}
